libpromises/evalfunction.c

```diff
@@ -6650,6 +6650,60 @@ static FnCallResult FnCallReadData(ARG_UNUSED EvalContext *ctx,
     return (FnCallResult) { FNCALL_SUCCESS, (Rval) { json, RVAL_TYPE_CONTAINER } };
 }
 
+static FnCallResult FnCallReadModuleProtocol(
+    ARG_UNUSED EvalContext *ctx,
+    ARG_UNUSED const Policy *policy,
+    const FnCall *fp,
+    const Rlist *args)
+{
+    assert(fp != NULL);
+
+    if (args == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function '%s' requires at least one argument", fp->name);
+        return FnFailure();
+    }
+
+    const char *input_path = RlistScalarValue(args);
+
+    char module_context[CF_BUFSIZE] = {0};
+
+    FILE *file = safe_fopen(input_path, "rt");
+    if (file == NULL)
+    {
+        return FnReturnContext(false);
+    }
+
+    StringSet *module_tags = StringSetNew();
+    long persistence = 0;
+
+    size_t line_size = CF_BUFSIZE;
+    char *line = xmalloc(line_size);
+
+    bool success = true;
+    for (;;)
+    {
+        const ssize_t res = CfReadLine(&line, &line_size, file);
```

Comment:
I like the `while (CfReadLine() != -1)` variant better.

Reply:
TBH, this was just copy pasted from the commands promise version, I didn't write this.